src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_search.cs

```diff
@@ -1,13 +1,38 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-
 namespace System.Buffers.Reader
 {
-    // TODO: the TryReadUntill methods are very inneficient. We need to fix that.
+    // TODO: the TryReadUntil methods are very inneficient. We need to fix that.
     public static partial class BufferReaderExtensions
     {
-        public static bool TryReadUntill(ref BufferReader reader, out ReadOnlySequence<byte> bytes, byte delimiter)
+        public static bool TryReadUntil(ref BufferReader reader, out ReadOnlySpan<byte> bytes, byte delimiter)
+        {
+            int idx = reader.UnreadSegment.IndexOf(delimiter);
+            if (idx != -1)
+            {
+                bytes = reader.UnreadSegment.Slice(0, idx);
+                reader.Advance(idx + 1);
+                return true;
+            }
+            else
+            {
+                bool result = TryReadUntil(ref reader, out ReadOnlySequence<byte> sequence, delimiter);
+
+                if (sequence.IsSingleSegment)
+                {
+                    bytes = sequence.First.Span;
+                }
+                else
+                {
+                    bytes = sequence.ToArray();
```

Comment:
Due to the stitch up of data that spans between segments, this ends up allocating. For instance, in the 4KB json payload, we allocated ~400 bytes (if we had the data split into 2 segments). The more segments we have, the more we have to stitch up and allocate.

Reply:
Using an out parameter that is a span forces you to do that...

Reply:
Span is how we expose the "value" (i.e. subsection of the payload) to the user as part of the JsonReader.

```C#
    public struct Utf8JsonReader {
        public Utf8JsonReader(ref ReadOnlySequence<byte> data);
        public Utf8JsonReader(ReadOnlySpan<byte> data);
        public JsonTokenType TokenType { get; }
        public ReadOnlySpan<byte> Value { get; }
        public JsonValueType ValueType { get; }
        public bool Read();
    }
```